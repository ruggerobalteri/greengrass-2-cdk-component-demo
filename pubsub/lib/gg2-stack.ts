import * as cdk from '@aws-cdk/core';
import * as fs from 'fs';
import * as gg from '@aws-cdk/aws-greengrassv2';
import * as s3 from "@aws-cdk/aws-s3"
import * as s3deploy from "@aws-cdk/aws-s3-deployment"
import * as s3asset from "@aws-cdk/aws-s3-assets"
import * as iam from "@aws-cdk/aws-iam"
import * as path from 'path';

export class Gg2Stack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const ggRoleArn = "arn:aws:iam::144932442613:role/GreengrassV2TokenExchangeRole"
    
    const ggBucket = new s3.Bucket(this, 'ggBucket');
    
    // const componentDeployment = new s3deploy.BucketDeployment(this, 'componentDeployment', {
    //   sources: [s3deploy.Source.asset('./lib/component/1.0.0')],
    //   destinationBucket: ggBucket,
    // });
    // componentDeployment.node.addDependency(ggBucket);
    const componentAsset = new s3asset.Asset(this, 'SampleAsset', {
      path: path.join(__dirname, './lib/component/1.0.0'),
    });
    componentAsset.node.addDependency(ggBucket)
    

    const ggBucketName = ggBucket.bucketName
    const componentRecipe = fs.readFileSync('./lib/component/recipe.yml', {
      encoding: 'utf-8'
    }).replace('#BUCKET',ggBucketName)


    const bucketAccessPolicy = new iam.Policy(this, "BucketAccessPolicy",{
      policyName: "GGV2ComponentArtifactPolicy"
    })
    const policyStatement = new iam.PolicyStatement({
      actions: ["s3:GetObject"],
      resources: [`arn:aws:s3:::${ggBucketName}/*`]
      }
    )
    bucketAccessPolicy.addStatements(policyStatement)

    const ggRole = iam.Role.fromRoleArn(this, 'Role', ggRoleArn, {
      // Set 'mutable' to 'false' to use the role as-is and prevent adding new
      // policies to it. The default is 'true', which means the role may be
      // modified as part of the deployment.
      mutable: true,
    });
    bucketAccessPolicy.attachToRole(ggRole)
    const pubsubComponent = new gg.CfnComponentVersion(this, "PubsubComponent", {
      inlineRecipe: componentRecipe
    })
    pubsubComponent.node.addDependency(componentDeployment);
  }
}
