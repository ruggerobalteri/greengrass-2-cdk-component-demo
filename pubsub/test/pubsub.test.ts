import {
  expect as expectCDK,
  matchTemplate,
  MatchStyle,
  SynthUtils,
} from "@aws-cdk/assert";
import * as cdk from "@aws-cdk/core";
import * as pubsub from "../lib/pubsub-stack";

test("Empty Stack", () => {
  const app = new cdk.App();
  // WHEN
  const stack = new pubsub.pubsubStack(app, "MyTestStack");
  // THEN
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
