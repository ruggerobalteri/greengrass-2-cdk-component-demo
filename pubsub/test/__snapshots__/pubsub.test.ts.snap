// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Empty Stack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454ArtifactHash38F4EC5B": Object {
      "Description": "Artifact hash for asset \\"8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454\\"",
      "Type": "String",
    },
    "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3BucketBC49F77C": Object {
      "Description": "S3 bucket for asset \\"8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454\\"",
      "Type": "String",
    },
    "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3VersionKey134A241F": Object {
      "Description": "S3 key for asset version \\"8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfArtifactHash85F58E48": Object {
      "Description": "Artifact hash for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C": Object {
      "Description": "S3 bucket for asset \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70": Object {
      "Description": "S3 key for asset version \\"c24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cf\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68ArtifactHashD9A515C3": Object {
      "Description": "Artifact hash for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7": Object {
      "Description": "S3 bucket for asset \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
    "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F": Object {
      "Description": "S3 key for asset version \\"e9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketAccessPolicyE2FA1F7F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:s3:::",
                    Object {
                      "Ref": "ggBucket6E31AE57",
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GGV2ComponentArtifactPolicy",
        "Roles": Array [
          "GreengrassV2TokenExchangeRole",
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3Bucket55EFA30C",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersc24b999656e4fe6c609c31bae56a1cf4717a405619c3aa6ba1bc686b8c2c86cfS3VersionKey60329B70",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Layers": Array [
          Object {
            "Ref": "componentDeploymentAwsCliLayerE64ABD62",
          },
        ],
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3BucketBC49F77C",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3BucketBC49F77C",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "ggBucket6E31AE57",
                    "Arn",
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "ggBucket6E31AE57",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PubsubComponent": Object {
      "DependsOn": Array [
        "componentDeploymentAwsCliLayerE64ABD62",
        "componentDeploymentCustomResourceFB3DF128",
      ],
      "Properties": Object {
        "InlineRecipe": Object {
          "Fn::Join": Array [
            "",
            Array [
              "---
RecipeFormatVersion: 2020-01-25
ComponentName: com.amazon.greengrass2.cdk.demo.pubsub
ComponentDescription: PubSub component demo for GGv2
ComponentPublisher: Amazon
ComponentVersion: \\"1.0.0\\"
ComponentDependencies:
  awsiot.greengrasscoreipc.client:
    VersionRequirement: \\">=0.0.0\\"
    DependencyType: HARD
  awsiot.greengrasscoreipc.model:
    VersionRequirement: \\">=0.0.0\\"
    DependencyType: HARD
Manifests:
  - Platform:
      os: linux
      architecture: all
    Lifecycle:
      Install:
        Skipif: onpath python
        Script: sudo apt-get install python
      Run: |
        ln -f -s -t . {artifacts:path}/*; python pubsub.py\\"
    Artifacts:
      - URI: s3://",
              Object {
                "Ref": "ggBucket6E31AE57",
              },
              "/pubsub.py
",
            ],
          ],
        },
      },
      "Type": "AWS::GreengrassV2::ComponentVersion",
    },
    "componentDeploymentAwsCliLayerE64ABD62": Object {
      "DependsOn": Array [
        "ggBucket6E31AE57",
      ],
      "Properties": Object {
        "Content": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3BucketAEADE8C7",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameterse9882ab123687399f934da0d45effe675ecc8ce13b40cb946f3e1d6141fe8d68S3VersionKeyE415415F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Description": "/opt/awscli/aws",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
    "componentDeploymentCustomResourceFB3DF128": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "ggBucket6E31AE57",
      ],
      "Properties": Object {
        "DestinationBucketName": Object {
          "Ref": "ggBucket6E31AE57",
        },
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3BucketBC49F77C",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3VersionKey134A241F",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters8cb9c9d3e8d7c6d5edabcb692807b7e44de75b5abfed0d95b60eed636ef2a454S3VersionKey134A241F",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "ggBucket6E31AE57": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
